lib/ansible/plugins/action/copy.py

```diff
@@ -423,44 +422,43 @@ def run(self, tmp=None, task_vars=None):
         remote_src = boolean(self._task.args.get('remote_src', False), strict=False)
         local_follow = boolean(self._task.args.get('local_follow', True), strict=False)
 
-        result['failed'] = True
+        # TODO: move to controller side argspec
+        msg = ''
         if not source and content is None:
-            result['msg'] = 'src (or content) is required'
+            msg = 'src (or content) is required'
         elif not dest:
-            result['msg'] = 'dest is required'
+            msg = 'dest is required'
         elif source and content is not None:
-            result['msg'] = 'src and content are mutually exclusive'
+            msg = 'src and content are mutually exclusive'
         elif content is not None and dest is not None and dest.endswith("/"):
-            result['msg'] = "can not use content with a dir as dest"
-        else:
-            del result['failed']
+            msg = "can not use content with a dir as dest"
 
-        if result.get('failed'):
-            return self._ensure_invocation(result)
+        if msg:
+            raise AnsibleActionFail(msg, result=result)
 
         # Define content_tempfile in case we set it after finding content populated.
         content_tempfile = None
 
         # If content is defined make a tmp file and write the content into it.
         if content is not None:
+            if isinstance(content, str):
+                source = content
+            else:
+                # If content is not a string we encode it to JSON
+                # to write it out.
+                source = json.dumps(source)
             try:
-                # If content comes to us as a dict it should be decoded json.
-                # We need to encode it back into a string to write it out.
-                if isinstance(content, dict) or isinstance(content, list):
-                    content_tempfile = self._create_content_tempfile(json.dumps(content))
-                else:
-                    content_tempfile = self._create_content_tempfile(content)
-                source = content_tempfile
-            except Exception as err:
-                result['failed'] = True
-                result['msg'] = "could not write content temp file: %s" % to_native(err)
-                return self._ensure_invocation(result)
+                content_tempfile = self._create_content_tempfile(source)
+            except (AttributeError, UnicodeError) as err:
+                raise AnsibleActionFail("Unable to encode content for writing to file", result=result) from err
+            except (IOError, OSError) as err:
```

Comment:
`IOError` merged into `OSError` over a decade ago, in Python 3.3:
```suggestion
            except OSError as err:
```